[CALCITE-3113] Equivalent MutableAggregates with different row types should match with each other.

jinxing64 opened the pull request:
In current code, below test case fails:
```
  @Test public void testAggregate1() {
    checkMaterialize(
        "select count(*) as c0 from \"emps\" group by \"empid\"",
        "select count(*) as c1 from \"emps\" group by \"empid\"");
  }

java.lang.AssertionError
    at org.apache.calcite.plan.SubstitutionVisitor.go(SubstitutionVisitor.java:504)
    at org.apache.calcite.plan.SubstitutionVisitor.go(SubstitutionVisitor.java:465)
    at org.apache.calcite.plan.MaterializedViewSubstitutionVisitor.go(MaterializedViewSubstitutionVisitor.java:56)
    at org.apache.calcite.plan.RelOptMaterializations.substitute(RelOptMaterializations.java:200)
    at org.apache.calcite.plan.RelOptMaterializations.useMaterializedViews(RelOptMaterializations.java:72)
    at org.apache.calcite.plan.volcano.VolcanoPlanner.registerMaterializations(VolcanoPlanner.java:347)
```
The reason is that `MutableRel` self-defined the identity mechanism. Obviously the materialization above should be used.

This PR proposes to do equivalence checking for row types but except for the field names.

I added two test cases in `MaterializationTest`, without this change, both of them fails.

rubenada commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -542,9 +556,7 @@ assert equalType(
-                assert result.result.rowType.equals(result.call.query.rowType)
-                    : Pair.of(result.result, result.call.query);
-                equivalents.put(result.result, result.call.query);
+                equivalents.put(result1, result.call.query);
```
I think that we could keep the assert, but using `result1` instead of `result.result`:
```
assert result1.rowType.equals(result.call.query.rowType)
                    : Pair.of(result1, result.call.query);
```

jinxing64 replied:
Hmm, this change only compensate the "top" target, but there's chance that child-node of target is equivalent with child-node of query, while they have different row-types. That's why I removed this this `assert`. If we keep it here, the test case I added will fail

jinxing64 replied:
After second thinking, I decided to add back the `assert` for safe check, but the check only happens when `targetDescendant==target`

rubenada replied:
Ok

rubenada commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -531,8 +531,22 @@ assert equalType(
+                  result1 = result.result;
+                }
+                attempted.add(new Replacement(result.call.query, result1));
+                MutableRel parent = result.call.query.replaceInParent(result1);
```
`parent` variable is never used, we could get rid of it.

jinxing64 replied:
Sure

julianhyde commented:
What do we do elsewhere (e.g. in RelOptRules)? Do we check whether row-types are the same modulo field names?

danny0405 commented:
> What do we do elsewhere (e.g. in RelOptRules)? Do we check whether row-types are the same modulo field names?

My concern is the same, in `RelOptPlanner`, we think 2 rels are equivalent only there rowType and digest are all the same with each other, for row type equals, we mean the field names are also need to be the same, should we sync this rule for MaterializedView ?

jinxing64 commented:
@julianhyde @danny0405 THX a lot for review !
In this change, when `targetDescendant==target`, I add a compensation `Project`, thus the final replaced has exactly the same row-type as before. 
When `targetDescendant != target`, I take two nodes with same digest but different row-types as equivalent, thus to achieve the matching, but this mismatch of row-types will not be visible in the final result.

danny0405 commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -542,14 +556,14 @@ assert equalType(
                   // Meanwhile we stop matching the descendants and jump
                   // to the next subtree in pre-order traversal.
+                  assert result1.rowType.equals(result.call.query.rowType)
+                      : Pair.of(result1, result.call.query);
```
This assert is not that necessary, because it is always true now.

jinxing64 replied:
OK， I will remove it.

danny0405 commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -531,8 +531,22 @@ assert equalType(
+                  result1 = MutableProject.of(
+                      result.call.query.rowType, result.result, projects);
+                } else {
+                  result1 = result.result;
```
Instead of new a MutableProject, can we just check the project list of the result row type ? If i'm not mistaken, only if the query and target have same row expressions but different alias can be matched, maybe we should only check the expression for alias.

jinxing64 replied:
Sorry, I don't quite get it ~ Do you mean below ?
1. No need to new a `MutableProject`, though the row type is different, we just let it go
2. Add a check/assert for the row type, we allow name mismatch, but other parts of row type should match with each other

danny0405 replied:
> Sorry, I don't quite get it ~ Do you mean below ?
> 
> 1. No need to new a `MutableProject`, though the row type is different, we just let it go
> 2. Add a check/assert for the row type, we allow name mismatch, but other parts of row type should match with each other

Yeah, that is just what i mean.

danny0405 commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -531,8 +531,22 @@ assert equalType(
+                  final RexBuilder rexBuilder = target.cluster.getRexBuilder();
+                  List<RexNode> projects = new ArrayList<>(
+                      rexBuilder.identityProjects(result.call.query.rowType));
+                  result1 = MutableProject.of(
```
Let's give the `result1` a more readable name, how about `equivalentResult` ?

jinxing64 replied:
Yes, that'd better.

danny0405 commented:
I have give some advices for my understanding. But i'm not sure if it is better or worse when we allow to match of materializedView with different row type. Because our planning rules only consider strict row type match for equivalent rel nodes. Maybe @jcamachor can give a better review.

jinxing64 commented:
THX, @danny0405   for your comments,  I'll update the patch soon.

jinxing64 commented:
@julianhyde 
THX for your comments from JIRA. I updated this patch. Please take another look when you have time. In current code, I extracted out a  `Boolean rowTypesAreEquivalent(MutableRel rel0, MutableRel rel1, Litmus litmus)` method for row type compatibility checking.

rubenada commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -593,6 +593,23 @@ assert equalType(
+  /**
+   * Equivalence checking for row types, but except for the field names.
+   */
+  private Boolean rowTypesAreEquivalent(
```
Shouldn't this method return `boolean` instead of `Boolean` ?

jinxing64 replied:
Sorry, updated :)

rubenada commented:
LGTM

jinxing64 commented:
THX @rubenada  for approving !

danny0405 commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -593,6 +593,23 @@ assert equalType(
+  private boolean rowTypesAreEquivalent(
+      MutableRel rel0, MutableRel rel1, Litmus litmus) {
+    // Validation checking for row type, but except for the field names.
+    assert rel0.rowType.getFieldCount() == rel1.rowType.getFieldCount()
```
This comment seems redundant, because there is already a java doc comment for the method.

danny0405 commented on core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:
```diff
@@ -593,6 +593,23 @@ assert equalType(
+  private boolean rowTypesAreEquivalent(
+      MutableRel rel0, MutableRel rel1, Litmus litmus) {
+    // Validation checking for row type, but except for the field names.
+    assert rel0.rowType.getFieldCount() == rel1.rowType.getFieldCount()
```
Why we still use this single assert and the other use the litmus ?

jinxing64 replied:
THX Danny,
I updated, please check~

jinxing64 commented:
@danny0405 
THX for merging !
